### ISSUE TITLE
JerryScript and IoT.js should use the same memory allocator

### ISSUE THREAD

**Issue by rtakacs** [Samsung/iotjs#211]
Currently JerryScript and IoT.js (libuv, http-parser) don't share their allocators, and they both have an own reserved space. However, if one space is full, there is no more memory, even if the other space is nearly unused. I think this is a problem, and a single space would be a better solution.

What do you think? Which one do you prefer: IoT.js should use Jerry allocator, or Jerry should use the system allocator?

**Comment by seanshpark** [comment]
@rtakacs , can you please explain more about your idea with the allocator? I mean with some example codes?

**Comment by ILyoan** [comment]
@rtakacs Agree with you for later option.

IoT.js uses 'new' or 'malloc' operation for dynamic allocation and the object will be generated in the different space from Jerry heap.

If there is API provided from JerryScript for allocating chunks of memory on JerryHeap, we can use the API.

**Comment by zherczeg** [comment]
@seanshpark currently we have two memory pools: one for Jerry and the other is for malloc. Let's assume we have 150K memory, and we assign 100K for Jerry, and 50K for malloc. In this case the http module cannot be loaded, since it requires 130K Jerry memory, although we have 50K unused memory. If Jerry would have all 150K memory, the module could be loaded. But in this case the other modules like libuv would not work, since malloc would fail. The proposal is to redirect all malloc to Jerry-heap (or all Jerry-heap to malloc) regardless its source.

**Comment by seanshpark** [comment]
@zherczeg , thank you for detailed explanations.
For small memory systems, we can benefit as you explained, for systems having bit more memory like can give Jerry full 256K and system has, say, 128K. For this case adding two doesn't seem to benefit. (If I'm not understand quite well, please point) So I have some question...
1) If we merge the two how can it be done? Something like  separate Jerry heap manager from Jerry so that libuv can also use.
2) Seems like mixing GC and manual alloc/free. Is this OK?
3) Can we separate them by configuration when needed, may be at build time? Do we need this?
4) Can Jerry heap size grow bigger than 256K? Any side effects?
5) How about performance? Is there any research or reports similar to this case?

**Comment by zherczeg** [comment]
1) SRR please correct me if I am wrong, but it seems to me that Jerry has a "heap" allocator, which manages all Jerry memory, and based on a classic alloc / free mechanism. We would redirect all malloc / free requests to this allocator (e.g. replacing the #include <malloc.h> with our own). This is not a trivial task if we don't want to change libuv. We could also propose patches to libuv to support custom allocators. We have multiple implementation options here.
2) GC just calls normal free when it frees something, so this has no effect on JavaScript object management
3) We could surely do this, but that would mean maintaining two code paths.
4) As far as I heard, it could grow to 512K. But that is the maximum. The side effect is performance, if Jerry allocator is less efficient than malloc, we will see a performance drop.
5) I cannot predict the effect on performance at this moment. We probably need to implement a majority of this work before we can evaluate that.

As a summary, this is not a trivial work, and it has risks (may not work at all). That is why we started a discussion before doing it. So the question is, shall we try it?

**Comment by lemmaa** [comment]
@zherczeg @seanshpark  Let's try. It will be a good experiment at least. I think it's a great chance of experiment compared to other applications using heavy JS engine.

**Comment by esevan** [comment]
I would like to try this, but I'm not really sure which direction of the solution is appropriate to begin. There might be two simple approaches for the solution of this problem. 

1) Using jerryScript heap for platform memory allocation
 In this case, is it able to span the jerry heap dynamically, so that js application could use whole system memory not using only statically allocated jerry heap?
2) Using malloc in jerryScript as ducktape does, so that jerryScript allocates the memory on demand
 Could we ignore malloc memory overhead in this case? Otherwise, how can we say IoT.js or jerryScript uses only XX Bytes?  (i.e. ptmalloc2 manages the memory and it's transparent to the user)

Very appreciate with Any idea of you.

**Comment by zherczeg** [comment]
I would go to direction 2. You need an allocator where you can specify the address range of the returned pointer:

malloc(size, addr_min, addr_max, align);

The returned pointer must be >=addr_min and < addr_max. The min would be the base for compressed pointers, and addr_max-addr_min <= 512K. Even the allocator could recommend the min. Jerry also need 8 byte aligned blocks. Furthermore Jerry can also tell the size of most blocks when the memory is freed. If your allocator is clever, it can use this information to use less memory. Interface:

free(ptr, size);

E.g. Jerry allocator uses zero bytes extra memory for allocated data.

**Comment by yichoi** [comment]
Closing.

About direction 1), jerry allocator will be in charge of managing iotjs and libtuv memory management, which will be unreasonable burden.

And about direction 2), as @zherczeg mentioned, below conditions would be important:

- Efficient 8-byte aligned block allocation
- Restricting returned address as addr_max-addr_min <= 512K
- Efficient alloc(size) / free(ptr, size), i.e., without extra metadata to save block size

Otherwise there will be big extra inefficiency on managing jerryscript heap.

Since none of our current [supported platforms](https://github.com/Samsung/iotjs/wiki/Getting-Started#supported-platforms) satisfies these conditions, I'm closing this issue now.

Please reopen it or create a new issue, if some port-specific requirement exists.

**Comment by esevan** [comment]
@yichoi About direction 2, I don't think we need to allocate the memory in unit of every allocation. If we expand the allocation granularity from the object to 8KB or 16KB and keep multiple memory bases instead of a single heap base, we can still use compressed pointer as well as reduce waste by over-provisoned heap.

Here is the current memory waste in several IoT.js module test codes when 256KB jerry heap is configured.
![image](https://cloud.githubusercontent.com/assets/8223765/20639364/da0acbfe-b404-11e6-9bb8-0fb7a9ec094b.png)

Moreover, I made [amazon echo](https://www.amazon.com/Amazon-Echo-Bluetooth-Speaker-with-WiFi-Alexa/dp/B00X4WHP5E)-like application in private for the evaluation. 
The application **_jerryscript runtime memory footprint_** in several environments is as below.
Original: 128KB jerry heap configured IoT.js (Always using 128KB system memory)
Ideal: Memory allocation and free are profiled in jerry internal.
Segment-allocation: 8KB on-demand memory allocation and free if 8KB segment is empty.

![image](https://cloud.githubusercontent.com/assets/8223765/20639389/7cc53c80-b405-11e6-83aa-de6b210ae7b2.png)

Now, the performance issue should be checked in segment-allocation, but what do you think of the memory consumption of jerryScript? If you agree with that the jerryScript memory consumption should be on-demand maintaining compressed pointer structure, I'll send PR of this modification version after some refactoring process.
